ann_benchmarks/algorithms/luceneknn.py

```diff
@@ -0,0 +1,124 @@
+        self.metric = metric
+        self.dimension = dimension
+        self.param = param
+        self.short_name = f"luceneknn-{param['M']}-{param['efConstruction']}"
+        self.simFunc = VectorSimilarityFunction.DOT_PRODUCT if self.metric == "angular" \
+            else VectorSimilarityFunction.EUCLIDEAN
+        if self.metric not in ("euclidean", "angular"):
+            raise NotImplementedError(f"Not implemented for metric {self.metric}")
+
+    def done(self):
+        if self.dir:
+            self.dir.close()
+
+    def fit(self, X):
+        if self.dimension != X.shape[1]:
+            raise Exception(f"Configured dimension {self.dimension} but data has shape {X.shape}")
+        if self.metric == 'angular':
+            X = sklearn.preprocessing.normalize(X, axis=1, norm='l2')
+        iwc = IndexWriterConfig().setOpenMode(IndexWriterConfig.OpenMode.CREATE)
+        codec = Codec(self.param['M'], self.param['efConstruction'])
+        iwc.setCodec(codec)
+        iwc.setRAMBufferSizeMB(1994.0)
+        self.dir = FSDirectory.open(Paths.get(self.short_name + ".index"))
+        iw = IndexWriter(self.dir, iwc)
+        fieldType = KnnVectorField.createFieldType(self.dimension, self.simFunc)
+        id = 0
+        # X is a numpy matrix, JArray casting only works on python lists.
+        X = X.tolist()
+        for x in X:
+            doc = Document()
+            doc.add(KnnVectorField("knn", JArray('float')(x), fieldType))
+            doc.add(StoredField("id", id))
+            id += 1
+            iw.addDocument(doc)
+            if id + 1 % 1000 == 0:
+                print(f"LuceneKNN: written {id} docs")
+        # Force merge so only one HNSW graph is searched.
+        iw.forceMerge(1)
+        print(f"LuceneKNN: written {id} docs")
+        iw.close()
+        self.searcher = IndexSearcher(DirectoryReader.open(self.dir))
+
+    def set_query_arguments(self, ef):
+        self.name = f"luceneknn-{self.dimension}-{self.param['M']}-{self.param['efConstruction']}-{ef}"
+        self.ef = ef
+
+    def query(self, q, n):
+        if self.metric == 'angular':
+            q = q / np.linalg.norm(q)
+        return self.run_knn_query(self.ef, n, q.tolist())
+
+    def prepare_batch_query(self, X, n):
+        if self.metric == 'angular':
+            X = sklearn.preprocessing.normalize(X, axis=1, norm='l2')
+        self.queries = X.tolist()
+        self.topK = n
+
+    def run_batch_query(self):
+        self.res = []
+        append = self.res.append
+        n = self.topK
+        num_candidates = self.ef
+        for q in self.queries:
+            append(self.run_knn_query(num_candidates=num_candidates, n=n, q=q))
+
+    def get_batch_results(self):
+        return self.res
+
+    def run_knn_query(self, num_candidates, n, q):
+        query = KnnVectorQuery("knn", JArray('float')(q), num_candidates)
+        topdocs = self.searcher.search(query, n)
+        return [d.doc for d in topdocs.scoreDocs]
```

Comment:
I just noticed -- this seems fragile because doc IDs aren't guaranteed to match the insertion order. For example, two segments could get merged together, which could totally reorder the document IDs. I noticed that you added an `id` stored field, maybe you intend to be using it here?

Reply:
🤦

Reply:
Actually, this isn't the only problem. I misunderstood PyLucene. Some more work is required. I need to write a pylucene Python extension point and hook into that...